addon/mixins/stateful.js

```diff
@@ -0,0 +1,299 @@
+    const actions = Object.assign({}, this.actions);
+    this._initializeState(actions, '_root', '_root');
+    this.actions._root = actions;
+  },
+
+  _defaultStateMapping: undefined,
+
+  init(...args) {
+    this._super(...args);
+    this._initializeStateHierarchy();
+    this.getStateTask('_root').perform(true);
+    const stateTaskNames = this.get('stateTaskNames') ;
+    Object.defineProperty(this, 'currentState', {
+      value: computed(...stateTaskNames.map(stateTask => `${stateTask}.isRunning`), this.getCurrentStateName),
+    });
+    // kick off the computed properties
+    this.get('state');
+    this.get('currentState');
+  },
+
+  send(actionName, ...args) {
+    let stateParts = this.get('currentState').split('.');
+    let shouldBubble = true;
+
+    while(shouldBubble && stateParts.length > 0) {
+      const action = this.get(`actions._root.${stateParts.join('.')}.${actionName}`);
+      if (action) {
+        shouldBubble = action.apply(this, args) === true;
+      }
+      stateParts.pop();
+    }
+    if (shouldBubble) {
+      const rootAction = this.actions._root[actionName];
+      if (!rootAction) {
+        throw new Error(ERRORS.NO_ROOT_ACTION(actionName, this));
+      }
+      rootAction.apply(this, args);
+    }
+  },
+
+  transitionTo(stateName) {
+    if (!this.checkIfStateExists(stateName)) {
+      throw new Error(ERRORS.NO_SUCH_STATE(stateName));
+    }
+    this.getStateTask(stateName).perform(true);
+  },
+
+  getStateTask(stateName) {
+    return this.get(getStateTaskName(stateName));
+  },
+
+  getDefaultStateName(stateName) {
+    return this._defaultStateMapping[stateName];
+  },
+
+  checkIfStateExists(stateName) {
+    const stateNames = this.get('stateNames');
+    return stateNames.includes(stateName);
+  },
+
+  *_stateTaskFunction(stateName, shouldStartDefaultSubtask = true) {
+    if (stateName !== '_root') {
+      const superStateName = getSuperStateName(stateName);
+      const superState = this.getStateTask(superStateName);
+      if (!superState.get('isRunning')) {
+        // during transitions into a deeply nested structure the superstates will not
+        // be running so kick them off first
+        superState.perform(false);
+      }
+    }
+
+    const stateActions = this.get(`actions.${stateName}`) || {};
+    try {
+      this.trigger(`try_${stateName}`);
+      yield executeStateHook(this, stateActions._try);
+      if (shouldStartDefaultSubtask) {
+        const defaultStateName = this.getDefaultStateName(stateName);
+        if (defaultStateName) {
+          const defaultStateTask = this.getStateTask(defaultStateName);
+          // we're cascading down until we hit a leaf state
+          defaultStateTask.perform(true);
+        }
+      }
+      // never ending yield to keep the task running
+      yield Ember.RSVP.defer().promise;
+    } catch(e) {
+      // TODO: bubble the exception up
+      if (stateActions._catch) {
+        stateActions._catch(e);
```

Comment:
this will have to be done through `executeStateHook` or similar, too, or else we loose the context (`this`).